File: frontend/src/types/k8s.ts
export interface ClusterStatus {
  phase: string;
  ready_replicas: number;
  total_replicas: number;
  coordinator_endpoint: string;
  message: string;
  pods: PodInfo[];
  volumes: VolumeInfo[];
  nodes: NodeInfo[];
  services: ServiceInfo[];
}

export interface PodInfo {
  name: string;
  role: string;
  status: string;
  cpu_usage_percent: number;
  memory_bytes: number;
  memory_limit_bytes: number;
  disk_usage_percent: number;
  restart_count: number;
  age_seconds: number;
}

export interface VolumeInfo {
  name: string;
  access_mode: string;
  capacity_bytes: number;
  used_bytes: number;
  bound_pod: string;
}

export interface ScaleRequest {
  replicas: number;
}

export interface KedaConfig {
  enabled: boolean;
  min_replicas: number;
  max_replicas: number;
}

export interface NodeInfo {
  name: string;
  status: string;
  roles: string[];
  cpu_capacity: number;
  memory_capacity_bytes: number;
  cpu_allocatable: number;
  memory_allocatable_bytes: number;
  pod_count: number;
  kubelet_version: string;
}

export interface ServiceInfo {
  name: string;
  type: string;
  cluster_ip: string;
  ports: ServicePort[];
}

export interface ServicePort {
  name: string;
  port: number;
  target_port: number;
  node_port?: number;
}

export interface K8sClusterEvent {
  type: string;
  reason: string;
  message: string;
  involved_object: string;
  timestamp: string;
}

export function podIsReady(pod: PodInfo): boolean {
  return pod.status === "Running";
}

export function podAgeDisplay(pod: PodInfo): string {
  const secs = pod.age_seconds;
  if (secs < 60) return `${secs}s`;
  if (secs < 3600) return `${Math.floor(secs / 60)}m`;
  if (secs < 86400) return `${Math.floor(secs / 3600)}h`;
  return `${Math.floor(secs / 86400)}d`;
}

export function podMemoryDisplay(pod: PodInfo): string {
  const mb = Math.floor(pod.memory_bytes / (1024 * 1024));
  if (mb < 1024) return `${mb} MB`;
  return `${(mb / 1024).toFixed(1)} GB`;
}

export function defaultClusterStatus(): ClusterStatus {
  return {
    phase: "",
    ready_replicas: 0,
    total_replicas: 0,
    coordinator_endpoint: "",
    message: "",
    pods: [],
    volumes: [],
    nodes: [],
    services: [],
  };
}

File: crates/megafactory-sql-types/src/k8s.rs
use serde::{Deserialize, Serialize};

/// Overall cluster status from MegaDB CRD.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ClusterStatus {
    pub phase: String,
    pub ready_replicas: i32,
    pub total_replicas: i32,
    pub coordinator_endpoint: String,
    pub message: String,
    pub pods: Vec<PodInfo>,
    pub volumes: Vec<VolumeInfo>,
}

/// Status of a single pod in the cluster.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PodInfo {
    pub name: String,
    /// "coordinator" or "worker"
    pub role: String,
    /// "Running", "Pending", "Failed", "Terminating"
    pub status: String,
    pub cpu_usage_percent: f64,
    pub memory_bytes: u64,
    pub memory_limit_bytes: u64,
    pub disk_usage_percent: f64,
    pub restart_count: i32,
    pub age_seconds: u64,
}

/// Status of a Persistent Volume Claim.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct VolumeInfo {
    pub name: String,
    /// "ReadWriteMany" or "ReadWriteOnce"
    pub access_mode: String,
    pub capacity_bytes: u64,
    pub used_bytes: u64,
    pub bound_pod: String,
}

/// Request to scale the cluster.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScaleRequest {
    pub replicas: i32,
}

/// KEDA autoscaling configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KedaConfig {
    pub enabled: bool,
    pub min_replicas: i32,
    pub max_replicas: i32,
}

impl PodInfo {
    pub fn is_ready(&self) -> bool {
        self.status == "Running"
    }

    pub fn age_display(&self) -> String {
        let secs = self.age_seconds;
        if secs < 60 {
            format!("{secs}s")
        } else if secs < 3600 {
            format!("{}m", secs / 60)
        } else if secs < 86400 {
            format!("{}h", secs / 3600)
        } else {
            format!("{}d", secs / 86400)
        }
    }

    pub fn memory_display(&self) -> String {
        let mb = self.memory_bytes / (1024 * 1024);
        if mb < 1024 {
            format!("{mb} MB")
        } else {
            format!("{:.1} GB", mb as f64 / 1024.0)
        }
    }
}
